packages/EasyCodingStandard/src/Console/Output/OutputFormatterFactory.php

```diff
@@ -0,0 +1,22 @@
+<?php declare(strict_types=1);
+
+namespace Symplify\EasyCodingStandard\Console\Output;
+
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symplify\EasyCodingStandard\Configuration\Option;
+use Symplify\EasyCodingStandard\Contract\Console\Output\OutputFormatterInterface;
+
+final class OutputFormatterFactory
+{
+    public static function create(ContainerInterface $container, InputInterface $input): OutputFormatterInterface
+    {
+        $formatOption = $input->getOption(Option::OUTPUT_FORMAT_OPTION);
+
+        if ($formatOption === Option::JSON_OUTPUT_FORMAT) {
+            return $container->get(JsonOutputFormatter::class);
+        }
+
+        return $container->get(TableOutputFormatter::class);
+    }
+}
```

Comment:
Instead of factory and container magic, use collector pattern:

```php
<?php

final class OutputFormatterCollector
{
    /**
     * @var OutputFormatterInterface[]
     */
    private $outputFormatters = [];
    
    /**
    * @param OutputFormatterInterface[]
    */
    public function __construct(array $outputFormatters)
    {
        foreach ($outputFormatters as $outputFormatter) {
            $this->outputFormatters[$outputFormatter->getName()] = $outputFormatter;
        }
    }
    
    public function getByName(string $name): OutputFormatterInterface
    {
        if (!isset($this->outputFormatters[$name])) {
            // exception: "not found, pick one of these:" . array_keys($this->outputFormatters) 
        }
        
        return $this->outputFormatters[$name];
    }
}
```

Reply:
Where can I read more about this pattern in this context? First time hearing about it.

Reply:
Do you use this pattern anywhere in the codebase? I don't quite understand where should `OutputFormatterCollector` be initialized.

Reply:
https://www.tomasvotruba.cz/clusters/#collector-pattern-the-shortcut-hack-to-solid-code

Reply:
> Do you use this pattern anywhere in the codebase? 

- https://github.com/rectorphp/rector/blob/f6f8b9d56861fda86f43d303eaa5f59e530f88ac/packages/YamlRector/src/YamlFileProcessor.php#L18
- https://github.com/rectorphp/rector/blob/f6f8b9d56861fda86f43d303eaa5f59e530f88ac/packages/FileSystemRector/src/FileSystemFileProcessor.php#L18
- https://github.com/rectorphp/rector/blob/aaf409eed279e71a5e33ca9b820f23213392105d/src/PhpParser/NodeTraverser/RectorNodeTraverser.php#L13

> I don't quite understand where should OutputFormatterCollector be initialized.

Just create the class like any other. It's loaded by Symfony Container

Reply:
Thanks! Now it should be easy.

At first, I was trying to use https://symfony.com/doc/current/service_container/expression_language.html to resolve which interface to use.

Then tried https://symfony.com/doc/current/service_container/synthetic_services.html.

Finally tried using factory inside `config.yml`, but could not get options inside the factory.

😓

Reply:
I wonder if this is ok, or should I keep everything in execute:

```php
final class CheckCommand extends Command
{
//...
    protected function initialize(InputInterface $input, OutputInterface $output): void
    {
        $outputFormat = $input->getOption(Option::OUTPUT_FORMAT_OPTION);
        $this->outputFormatter = $this->outputFormatterCollector->getByName($outputFormat);
    }

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $this->ensureSomeCheckersAreRegistered();

        $this->configuration->resolveFromInput($input);
        $processedFilesCount = $this->ecsApplication->run();

        return $this->outputFormatter->report($processedFilesCount);
    }
//...
}
```

Reply:
Put it into one method `execute()` please, it's pretty rare to use anything else.